### ISSUE TITLE
Support type inference for fold, method and array

### ISSUE THREAD

**Issue by farodin91** [intellij-rust/intellij-rust#1603]
example: fold 
```rust
let a = vec![1, 2, 3];

// the sum of all of the elements of a
let sum = a.iter().fold(0, |acc/*: {int} (should be)*/, &x| acc + x);

assert_eq!(sum, 6);
```

example: method
```rust
let a = vec!["1", "2", "4"];

// the sum of all of the elements of a
let as_strings = a.iter().map(ToString::to_string);
    //^ Map<Iter<&str>, ToString::to_string> (should be)
```

example: array

```rust
let a = ["1", "2", "4"];

// the sum of all of the elements of a
let iter = a.iter();
    //^ Iter<&str> (should be)
```

**Comment by vlad20012** [comment]
Hmm, fold currently works for me:
```rust
fn main() {
    let vec: Vec<i32> = Vec::new();
    let sum = vec.iter().fold(0, |acc, &x| acc + x);
    sum
    //^ i32
}
```

**Comment by farodin91** [comment]
@vlad20012 Work hint to ?

**Comment by vlad20012** [comment]
@farodin91 Type hinths? Heh, I don't see any type hints on the latest nightly at all:
![image](https://user-images.githubusercontent.com/3221931/29241914-e62b121a-7f8b-11e7-966b-1860c569fcb4.png)

Previously I've disabled hints due to #1548 and did not enable them until now. IDE restart did not help. Nice.

Anyway, `fold` type inference works fine.

**Comment by farodin91** [comment]
Inner type inference in lambda didn't work for me acc is B not i32.

**Comment by vlad20012** [comment]
@farodin91 Uh, now understand. Your example was too ambigous =)
Yes, `acc` inference faild. Ok

**Comment by vlad20012** [comment]
"array" issue was fixed by #1664
"method" will be fixed by #1767
"fold" seems to work now, but I don't completely now what commit fixed it
